### PR TITLE
Add uniqueConstraintExists precondition (#1881)

### PR DESCRIPTION
**Pull request by crizzis**
## Environment

**Liquibase Version**: 

**Liquibase Integration & Version**: <Pick one: CLI, maven, gradle, spring boot, servlet, etc.>

**Liquibase Extension(s) & Version**: 

**Database Vendor & Version**: all (tested on Postgres)

**Operating System Type & Version**: all (tested on Mac OS 10.15.7)

## Pull Request Type

- [ ] Bug fix (non-breaking change which fixes an issue.)
- [x] Enhancement/New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description

**Add uniqueConstraintExists precondition**

This PR
- adds a new XSD schema element, `<uniqueConstraintExists>`
- adds code support for that element in the form of the `UniqueConstraintExistsPrecondition` class
- fixes the issue which caused unique constraints not being returned in a snapshot, initially causing the new functionality to not work properly

## Steps To Reproduce

- Add a new changeset with a precondition like: 
```xml
<preConditions>
            <uniqueConstraintExists constraintName="..." tableName="..." columnNames="..." />
        </preConditions>
```
- Observe that the changset is applied if the unique constraint defined by the precondition exists
- Observe that the changset is **NOT** applied if the unique constraint defined by the precondition does not exist

## Actual Behavior
Before there was no support whatsoever for checking the existence of a unique constraint

## Expected/Desired Behavior
Now, changesets can be conditioned on whether or not a given unique constraint exists

## Additional Context

I added a new XSD version, wasn't sure if I was supposed to do that or extend the existing one, since this is not a breaking change

## Fast Track PR Acceptance Checklist:

- [x] Build is successful and all new and existing tests pass
- [ ] Added [Unit Test(s)](https://liquibase.jira.com/wiki/spaces/LB/pages/1274937609/How+to+Write+Liquibase+Core+Unit+Tests)
- [x] Added [Integration Test(s)](https://liquibase.jira.com/wiki/spaces/LB/pages/1276608569/How+to+Write+Liquibase+Core+Integration+Tests)
- [ ] Added [Test Harness Test(s)](https://github.com/liquibase/liquibase-test-harness/pulls)
- [ ] Documentation Updated

## Need Help?
Come chat with us on our [discord channel](https://discord.com/channels/700506481111597066/700506481572839505)

### CODE REVIEW

File: liquibase-core/src/main/java/liquibase/snapshot/jvm/UniqueConstraintSnapshotGenerator.java [hunk 1]
```diff
@@ -176,24 +177,34 @@ protected void addTo(DatabaseObject foundObject, DatabaseSnapshot snapshot) thro
...
+                List<String> conditions = new ArrayList<>();
                 sql = "select const.CONSTRAINT_NAME, COLUMN_NAME, const.constraint_schema as CONSTRAINT_CONTAINER "
                         + "from " + database.getSystemSchema() + ".table_constraints const "
                         + "join " + database.getSystemSchema() + ".key_column_usage col "
                         + "on const.constraint_schema=col.constraint_schema "
                         + "and const.table_name=col.table_name "
-                        + "and const.constraint_name=col.constraint_name "
-                        + "where const.constraint_catalog='" + database.correctObjectName(schema.getCatalogName(), Catalog.class) + "' ";
+                        + "and const.constraint_name=col.constraint_name ";
+                if (schema.getCatalogName() != null) {
```
Comment (crizzis) [inline liquibase-core/src/main/java/liquibase/snapshot/jvm/UniqueConstraintSnapshotGenerator.java:187]: Two problems solved here: 
* When catalog name and schema name were missing, the query would try to look for a catalog/schema with the literal name `'null'`
* Missing space between `const.constraint_name=XXX` and `order by`

File: liquibase-integration-tests/src/test/resources/liquibase/extension/testing/command/history.test.groovy [hunk 2]
```diff
@@ -61,15 +61,11 @@ Optional Args:
...
-                Pattern.compile("""
-- Database updated at \\d+/\\d+.+. Applied 4 changeset\\(s\\) in \\d+.\\d+s, DeploymentId: \\d+
-  db/changelog/db.changelog-master.xml::1::nvoxland
-  db/changelog/sql/create_test2.sql::raw::includeAll
-  db/changelog/sql/create_test3.sql::raw::includeAll
-  db/changelog/changelog-x.xml::1571079854679-2::nathan \\(generated\\)
-""".replace("\r", "").trim())
-        ]
-
+~/- Database updated at .+\. Applied 4 changeset\(s\) in \d+.\d+s, DeploymentId: \d+
```
Comment (crizzis) [inline liquibase-integration-tests/src/test/resources/liquibase/extension/testing/command/history.test.groovy:64]: Turns out the date pattern was locale sensitive, relaxed the regex a little

File: liquibase-core/src/main/resources/www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.8.xsd [hunk 3]
```diff
@@ -0,0 +1,1400 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (nvoxland) [reply liquibase-core/src/main/resources/www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.8.xsd:1]: I used this PR as a chance to try a new strategy of creating a dbchangelog-NEXT.xsd as a target. I moved your changes to that.

**Review by nvoxland — APPROVED** [review]
The changes look good, and I greatly appreciate your comments within the PR

**Comment by nvoxland** [comment]
Code review and testing notes:

- Removed tests from IntegrationTest classes was due to moving that test to be ran everywhere and so not needing to be re-defined
- Removed test from AbstractIntegrationTest was because it was just re-running a call that's already called in many other tests so no need to re-test it
- I did not try to run the changes
- The new uniqueConstraintExists precondition works across all databases
- There is code cleanup/fixing around snapshotting unique constraints which only impacts postgresql & cockroachdb

**Review by suryaaki2 — APPROVED** [review]

**Comment by nvoxland** [comment]
In testing this, we found some issues when tableName was not set. There was also an overall problem where we weren't doing anything with the "validate" function.

I pushed a change that makes tableName required, since some databases will need that because constraint names are only unique within a table, and the current UniqueConstraint object model needs the table to know the schema of the constraint.

I also pushed a change to use the validation errors.

**Comment by nvoxland** [comment]
I also took the warn() logic out, since in theory you can include both in order to check that the constraint name on the columns exist.

**Comment by EagleInNight** [comment]
Faced a problem that uniqueConstraintExists precondition fails event if constraint exists. At least for mysql

**Comment by EagleInNight** [comment]
So, it couldn't be find because of 'null' catalogName (UniqueConstraintSnapshotGenerator.java:173)
Currently fixed issue with specifying both schemaName and catalogName in uniqueConstraintExists precondition

**Pull request merged**